kube-hetzner/terraform-hcloud-kube-hetzner
# Add letsencrypt clusterissuers + hetzner webhook + reflector + auto wildcard cert generation

# Add LetsEncrypt ClusterIssuers + Hetzner webhook + Reflector + auto wildcard cert generation

This PR adds LetsEncrypt ClusterIssuers Helm Chart installation so that the cluster is set up and ready to generate TLS certs by passing both `HTTP-01` and `DNS-01` ACME challenges. While the former requires no additional dependencies apart from the ClusterIssuer itself, the latter requires the [Hetzner API Webhook for the Cert Manager](https://github.com/vadimkim/cert-manager-webhook-hetzner), which is also installed and set up by another Helm Chart included in `cluster_issuers.yaml.tpl`. 

I've created new `var.hetzner_dns_api_token` to set up Hetzner DNS API Token in there (Hetzner delegated DNS Zone is a must for this to work), and that variable is then encoded and passed on to `cluster_issuers.yaml.tpl` to generate `hetzner-dns-secret`, which the webhook uses to pass the challenge.

Finally, I've included a `wildcard_cert.yaml.tpl` which triggers auto wildcard cert secret generation using `common_name`, a template variable obtained from regexing the `var.base_domain` FQDN to obtain the root domain and request `*.example.com`, and flagged Reflector annotations for the secret. [Reflector](https://github.com/emberstack/kubernetes-reflector/tree/main) is a little utility that replicates annotated secrets in all or specifically selected namespaces. This is especially useful for TLS certs since they are often needed in more than one namespace. Both the ClusterIssuers and the Reflector installation are obviously flagged through `var.enable_cluster_issuers` and `var.enable_reflector` and disabled by default.

## How to test
Assuming your DNS Zone is delegated to Hetzner:

1. Set `var.enable_cluster_issuers` and `var.enable_reflector` to `true`
2. Make sure your `var.base_domain` is duly populated within a valid FQDN and point its A record to your server (for single-node clusters) or cloud load balancer if you are using one
3. Make sure you've generated a Hetzner DNS Zone API Token and set it to `var.hetzner_dns_api_token`
4. Launch `tofu apply` or `terraform apply` to create your cluster

If everything is correct, whenever your cluster is ready you should see both ClusterIssuers (prod and staging) up and running.
* Run `kubectl get clusterissuers -n cert-manager --kubeconfig k3s_kubeconfig.yaml`

![image](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/4315171/65f59347-dd3f-450f-a415-a29202272d5b)

* Run `kubectl get certs -n cert-manager --kubeconfig k3s_kubeconfig.yaml`

![image](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/4315171/d4bcc28a-1e86-4742-9942-c4915999903a)

* Inspect the cert by `kubectl describe certs/wildcard-domain -n cert-manager --kubeconfig k3s_kubeconfig.yaml`

![image](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/4315171/c678b05f-84ab-46de-b245-17ba84ca35b8)

* Inspect other namespaces to check the correct secret replication by Reflector: `kubectl get secrets --all-namespaces --kubeconfig k3s_kubeconfig.yaml`

![image](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/assets/4315171/f4d091a8-950d-4aa0-991c-61940e8392d2)

## Caveats

* If no Hetzner DNS Zone API Token is set or it is wrong, LetsEncrypt ClusterIssuers will still be properly installed and you will be able to pass `http-01` challenges to generate FQDN domains but `dns-01` challenge will fail and no wildcard cert will be generated. Same applies if your DNS Zone is not delegated to Hetzner.

* You can disable Reflector if you don't need secret replication and it will only affect that, ClusterIssuers will still work as intended.

* Bear in mind that `dns-01` challenge needs to insert and propagate a record in the DNS Zone, so it can take up to 5 minutes to properly validate the wildcard cert. Everything is going as intended as long as the challenge logs do not return any errors.

**Att: Repo maintainers:** Please feel free to modify, comment, suggest, whatever you wish to do with this PR. Everything will be most welcome ;)
